### ISSUE TITLE
Fine-tune elmo model

### ISSUE THREAD

**Issue by paulomann** [allenai/allennlp#2904]
**System (please complete the following information):**
 - OS: Linux Ubuntu 18
 - Python version: 3.6.5
 - AllenNLP version: 0.8.3
 - PyTorch version: 1.1.0

**Question**

**I.** I have used a pretrained elmo biLM from allennlp's website, the Portuguese one. I loaded it with the `allennlp.modules.elmo.Elmo` class to train the linear combination parameters such that it gives better results for my downstream task.
**II.** Now I want to fine-tune with some specific data set to see what happens in my downstream task.
**III.** However, I tried the `allennlp fine-tune` command without success, since it seems that the command expects a gzip file, the model archive, which is something I don't have at the moment, only the hdf5 weights file (elmo_pt_weights.hdf5), and the options json, both provided on the Allennlp website.
**IV.** It's desirable to fine-tune the biLM with some text data from other domain (tweets), not the corpus used for my downstream task. That's why I can't just fine-tune the `allennlp.modules.elmo.Elmo` while training it for my downstream task.

Given this settings, I have 2 questoins:
**1.** Can I fine-tune it, if yes, where?
**2.** How should be the format of my data set? Should I have a vocabulary for it?

**TL;DR**: I want to fine-tune a pretrained model on Potuguese data with Portuguese tweets that I have. After that, I want to use this fine-tuned model in my downstream task. How can I fine-tune this pretrained model with the tweets that I have?

**Comment by kernelmachine** [comment]
The LSTM ELMo (with hdf5 weights) can be fine-tuned with the corresponding code: https://github.com/allenai/bilm-tf. There are instructions in the FAQ about fine-tuning to additional unlabeled data, as well as information about vocabulary and dataset format.

If you'd like to use `allennlp fine-tune`, your best bet currently is to use the Transformer ELMo (https://github.com/allenai/allennlp/blob/master/tutorials/how_to/training_transformer_elmo.md). A pretrained Transformer ELMo can be found here: s3://allennlp/models/transformer-elmo-2019.01.10.tar.gz.

**Comment by paulomann** [comment]
Hello @kernelmachine, thanks for your reply.

To the best of my knowledge, reading the bilm-tf repository, I still can't use it for my case, since it expects a .ckpt file, which I don't have because it's not available with the Portuguese pretrained model in the allennlp website.

I can't use the `allennlp fine-tune` approach either, because I have no `archive_file` from the pretrained Portuguese corpus, and therefore, I can't use the Transformer ELMo because it expects the `archive_file` as well.

Last, the .tar.gz you linked is the archive_file for Portuguese data set?

**Comment by brendan-ai2** [comment]
Hi @paulomann, this should be possible but will require a fair amount of hacking and glue code. If you really need to do it, I'd suggest something like the following:

1. Build a language model, as an `allennlp.models.model.Model`, that wraps [`allennlp.modules.elmo._ElmoBiLm`](https://github.com/allenai/allennlp/blob/master/allennlp/modules/elmo.py#L494). Principally you'll need to calculate the loss in the forward method. You should be able to take inspiration from [here](https://github.com/allenai/allennlp/blob/master/allennlp/models/language_model.py#L157). The options and weights files should be set to the pre-trained Portuguese ELMo.

2. Train this model normally using `allennlp train` on your tweet data. This is really fine tuning. As far as config, I think you'll be able to use [this dataset reader](https://github.com/allenai/allennlp/blob/master/allennlp/data/dataset_readers/simple_language_modeling.py) and AllenNLP will create a vocabulary automatically as normal. (You may wish to tweak the settings.) One thing to consider is dataset size. If you have a large amount of data, you'll need to deal with scaling issues like using multiple GPUs and sharding your data. This can get a bit complicated, so I'd avoid it if you can. Otherwise the transformer ELMo tutorial that @kernelmachine linked can perhaps serve as a guide.

3. Finally, in your downstream model:
    1. In your `__init__` use [`load_archive`](https://github.com/allenai/allennlp/blob/master/allennlp/models/archival.py#L150) to load the language model you created above.
    2. Pull out the (now) fine-tuned `_ElmoBiLm` and pass it to `allennlp.modules.elmo.Elmo`'s constructor as the `module`. Be sure to set `options_file`  and `weight_file` to `None`.
    3. Use that `Elmo` module normally in your downstream model.

I hope that helps. Good luck!

**Comment by paulomann** [comment]
Hello @brendan-ai2, 

Really thanks for the detailed answer, it surely helped me a lot. I contacted the authors of the Portuguese model, asking if they could send me the checkpoints. Although it is easier with the checkpoints, I can surely learn a lot more doing this approach you suggested, and I probably should be doing it anyway, in order to learn.

I appreciate every help you and your team gave me!

Thanks a lot!

**Comment by danieldominguete** [comment]
Hello @paulomann ! 

I´m too very interested about the checkpoints files from Portuguese Model. If it is possible, I would like to receiving them. Thanks in advance!

**Comment by brendan-ai2** [comment]
Very glad to be of help, @paulomann. Sounds like this issue can be closed, but feel free to comment again if anything comes up. Regards!

**Comment by saadi143** [comment]
@brendan-ai2 Can you please help me. I need your suggestion about ELMO embedding. I am trying to get the Elmo embedding with 300 or 256 dimensions. I got the pre-trained weight file from {[AllenNLP](https://allennlp.org/elmo). When I pass the weight and options file to Elmo function it takes time to train. 
Can you tell me what kind of training it has done. 
1. Is it fine-tuned?
2. Is it training from scratch?

And another thing is ->  can I Get Elmo embedding of dimension 300 from TensorFlow hub?

**Comment by brendan-ai2** [comment]
@saadi143, a detailed description of the training of ELMo is available in the paper: https://arxiv.org/pdf/1802.05365.pdf. The output dimensions on the site you linked are (I believe) all that we've trained, i.e. I don't think we have one with output size 300. Also, a small point, could you please open a new issue for any followup questions? It helps us triage them more effectively. Thanks!
